File: apps/api/src/lib/holdings.ts
/**
 * Investment holding utilities.
 * Adjusts InvestmentHolding records when trade transactions are created or deleted.
 */
import { prisma } from '@budget-tracker/db';
import type { DbClient } from './lifecycle/db-client.js';

interface TradeMetadata {
  direction: 'BUY' | 'SELL';
  assetType: 'Stock' | 'Bitcoin';
  ticker?: string;
  unitPrice: number;
  quantity: number;
  bitcoinUnit?: 'Bitcoin' | 'Sats';
  custodianId?: string;
  walletId?: string;
}

const detailNum = (v: number | { toNumber(): number }): number =>
  typeof v === 'number' ? v : v.toNumber();

/** Adapt a persisted TradeDetail row to the trade-holding input shape. */
export function tradeDetailToMetadata(d: {
  direction: string;
  assetType: string;
  ticker: string | null;
  quantity: number | { toNumber(): number };
  unitPrice: number | { toNumber(): number };
  bitcoinUnit: string | null;
  custodianId: string | null;
  walletId: string | null;
}): TradeMetadata {
  return {
    direction: d.direction as 'BUY' | 'SELL',
    assetType: d.assetType as 'Stock' | 'Bitcoin',
    ticker: d.ticker ?? undefined,
    unitPrice: detailNum(d.unitPrice),
    quantity: detailNum(d.quantity),
    bitcoinUnit: (d.bitcoinUnit ?? undefined) as 'Bitcoin' | 'Sats' | undefined,
    custodianId: d.custodianId ?? undefined,
    walletId: d.walletId ?? undefined,
  };
}

/**
 * Apply a trade's effect on investment holdings.
 *
 * @param tradeMetadata - The parsed trade metadata from the transaction
 * @param usdAmount - The USD amount of the trade
 * @param multiplier - 1 to apply (create), -1 to reverse (delete)
 */
export async function applyTradeToHolding(
  tradeMetadata: TradeMetadata,
  usdAmount: number,
  multiplier: 1 | -1 = 1,
  db: DbClient = prisma,
): Promise<{ costBasisAllocated?: number }> {
  const { direction, assetType, quantity: rawQuantity } = tradeMetadata;

  // Convert sats to BTC if needed
  const quantity =
    assetType === 'Bitcoin' && tradeMetadata.bitcoinUnit === 'Sats'
      ? rawQuantity / 100_000_000
      : rawQuantity;

  // Determine holding type and ticker
  const holdingType = assetType === 'Stock' ? 'STOCK' : 'BITCOIN';
  const ticker = assetType === 'Stock' ? tradeMetadata.ticker! : null;

  // Build match criteria using FK instead of accountName
  const matchWhere = {
    type: holdingType as 'STOCK' | 'BITCOIN',
    ...(assetType === 'Stock'
      ? { custodianId: tradeMetadata.custodianId! }
      : { walletId: tradeMetadata.walletId! }),
    ...(ticker ? { ticker } : { ticker: null }),
  };

  // Find existing holding
  const existing = await db.investmentHolding.findFirst({ where: matchWhere });

  // Determine effective direction considering multiplier
  // multiplier=-1 reverses: a BUY reversal acts like SELL, a SELL reversal acts like BUY
  const isBuy =
    (direction === 'BUY' && multiplier === 1) || (direction === 'SELL' && multiplier === -1);

  if (isBuy) {
    // Increment quantity and cost basis
    if (existing) {
      await db.investmentHolding.update({
        where: { id: existing.id },
        data: {
          quantity: { increment: quantity },
          costBasis: { increment: usdAmount },
        },
      });
    } else {
      // Look up custodian/wallet name for the holding display name
      let entityName: string;
      if (assetType === 'Stock') {
        const custodian = await db.custodian.findUniqueOrThrow({
          where: { id: tradeMetadata.custodianId! },
        });
        entityName = custodian.name;
      } else {
        const wallet = await db.wallet.findUniqueOrThrow({
          where: { id: tradeMetadata.walletId! },
        });
        entityName = wallet.name;
      }

      await db.investmentHolding.create({
        data: {
          name: entityName,
          ticker,
          type: holdingType as 'STOCK' | 'BITCOIN',
          quantity,
          costBasis: usdAmount,
          ...(assetType === 'Stock'
            ? { custodianId: tradeMetadata.custodianId! }
            : { walletId: tradeMetadata.walletId! }),
        },
      });
    }
    return {};
  } else {
    // Decrement quantity and proportional cost basis
    if (existing) {
      const currentQty = Number(existing.quantity);
      const currentCostBasis = Number(existing.costBasis ?? 0);

      // Proportional cost basis: (quantity sold / current quantity) * current cost basis
      const proportion = currentQty > 0 ? quantity / currentQty : 0;
      const costBasisReduction = currentCostBasis * proportion;

      await db.investmentHolding.update({
        where: { id: existing.id },
        data: {
          quantity: { decrement: quantity },
          costBasis: { decrement: costBasisReduction },
        },
      });

      return { costBasisAllocated: costBasisReduction };
    }
    // If no existing holding for a sell, nothing to decrement (shouldn't happen with proper validation)
    return {};
  }
}

// ─── Bitcoin Payment Holdings ───

interface BitcoinPaymentMetadata {
  walletId: string;
  quantity: number;
  bitcoinUnit: 'Bitcoin' | 'Sats';
  unitPrice: number;
}

/** Adapt a persisted BitcoinPaymentDetail row to the bitcoin-holding input shape. */
export function bitcoinDetailToMetadata(d: {
  walletId: string;
  quantity: number | { toNumber(): number };
  unitPrice: number | { toNumber(): number };
  bitcoinUnit: string;
}): BitcoinPaymentMetadata {
  return {
    walletId: d.walletId,
    quantity: detailNum(d.quantity),
    bitcoinUnit: d.bitcoinUnit as 'Bitcoin' | 'Sats',
    unitPrice: detailNum(d.unitPrice),
  };
}

/**
 * Compute the USD equivalent for a Bitcoin payment.
 *
 * - Bitcoin: usdAmount = quantity × unitPrice
 * - Sats:    usdAmount = (quantity / 100_000_000) × unitPrice
 */
export function computeUsdAmount(
  quantity: number,
  bitcoinUnit: 'Bitcoin' | 'Sats',
  unitPrice: number,
): number {
  const btcQuantity = bitcoinUnit === 'Sats' ? quantity / 100_000_000 : quantity;
  return btcQuantity * unitPrice;
}

/**
 * Back-calculate the unit price (USD per BTC) given a USD equivalent and quantity.
 *
 * This is the inverse of `computeUsdAmount`:
 *   computeUsdAmount(quantity, unit, backCalculateUnitPrice(usd, quantity, unit)) ≈ usd
 *
 * - If bitcoinUnit is 'Sats', converts quantity to BTC first.
 * - Returns usdEquivalent / btcQuantity.
 */
export function backCalculateUnitPrice(
  usdEquivalent: number,
  quantity: number,
  bitcoinUnit: 'Bitcoin' | 'Sats',
): number {
  const btcQuantity = bitcoinUnit === 'Sats' ? quantity / 100_000_000 : quantity;
  return usdEquivalent / btcQuantity;
}

/**
 * Apply a Bitcoin payment's effect on investment holdings.
 *
 * EXPENSE+multiplier=1 → decrement (spending BTC)
 * EXPENSE+multiplier=-1 → increment (reversing a spend)
 * INCOME/REFUND+multiplier=1 → increment (receiving BTC)
 * INCOME/REFUND+multiplier=-1 → decrement proportionally (reversing a receive)
 *
 * @param meta - The parsed Bitcoin payment metadata from the transaction
 * @param txType - The transaction type (EXPENSE, INCOME, or REFUND)
 * @param usdAmount - The USD equivalent amount of the transaction
 * @param multiplier - 1 to apply (create), -1 to reverse (delete)
 */
export async function applyBitcoinToHolding(
  meta: BitcoinPaymentMetadata,
  txType: 'EXPENSE' | 'INCOME' | 'REFUND',
  usdAmount: number,
  multiplier: 1 | -1 = 1,
  db: DbClient = prisma,
): Promise<void> {
  // Convert sats to BTC if needed
  const btcQuantity = meta.bitcoinUnit === 'Sats' ? meta.quantity / 100_000_000 : meta.quantity;

  // Find existing holding for this wallet
  const existing = await db.investmentHolding.findFirst({
    where: { type: 'BITCOIN', walletId: meta.walletId, ticker: null },
  });

  // Determine effective direction:
  // EXPENSE+1 = decrement, EXPENSE-1 = increment
  // INCOME/REFUND+1 = increment, INCOME/REFUND-1 = decrement
  const isIncrement =
    (txType === 'EXPENSE' && multiplier === -1) ||
    ((txType === 'INCOME' || txType === 'REFUND') && multiplier === 1);

  if (isIncrement) {
    if (existing) {
      await db.investmentHolding.update({
        where: { id: existing.id },
        data: {
          quantity: { increment: btcQuantity },
          costBasis: { increment: usdAmount },
        },
      });
    } else {
      // Auto-create holding — look up wallet name
      const wallet = await db.wallet.findUniqueOrThrow({
        where: { id: meta.walletId },
      });

      await db.investmentHolding.create({
        data: {
          name: wallet.name,
          type: 'BITCOIN',
          ticker: null,
          quantity: btcQuantity,
          costBasis: usdAmount,
          walletId: meta.walletId,
        },
      });
    }
  } else {
    // Decrement quantity and proportional costBasis
    if (existing) {
      const currentQty = Number(existing.quantity);
      const currentCostBasis = Number(existing.costBasis ?? 0);

      const proportion = currentQty > 0 ? btcQuantity / currentQty : 0;
      const costBasisReduction = currentCostBasis * proportion;

      await db.investmentHolding.update({
        where: { id: existing.id },
        data: {
          quantity: { decrement: btcQuantity },
          costBasis: { decrement: costBasisReduction },
        },
      });
    }
    // If no existing holding for a decrement, nothing to do (shouldn't happen with proper validation)
  }
}

File: rust/db/src/holdings.rs
//! Investment holdings — port of `applyTradeToHolding` from
//! `apps/api/src/lib/holdings.ts` (trade-holding hook, priority 10).
//!
//! **This is where ADR-033's two representations meet**, and the reason the
//! decision was to have two rather than one:
//!
//! - `InvestmentHolding.quantity` is a QUANTITY — TEXT holding an exact decimal.
//!   BTC needs 8 places and `InvestmentSnapshot.quantity` reaches 20 in
//!   production; cents would round `0.00000001` BTC to zero. It is added and
//!   subtracted HERE, in Rust, never in SQL: SQLite coerces TEXT to float on
//!   arithmetic, which would silently reintroduce the drift the representation
//!   exists to avoid.
//! - `InvestmentHolding.costBasis` is MONEY — INTEGER cents, exact under
//!   addition.
//!
//! The proportional cost basis on a sell is the one division in this module,
//! and per ADR-033's surviving rule it states its policy: round half away from
//! zero to the cent, with the residual left in the holding rather than
//! distributed, because the holding is the running figure and the allocation is
//! derived from it.

use anyhow::{Context, Result};
use avoir_core::money::Cents;
use rust_decimal::prelude::*;
use rust_decimal::{Decimal, RoundingStrategy};
use sqlx::SqliteConnection;

const SATS_PER_BTC: i64 = 100_000_000;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Direction {
    Buy,
    Sell,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AssetType {
    Stock,
    Bitcoin,
}

/// The trade as stored in `TradeDetail`.
#[derive(Debug, Clone)]
pub struct Trade {
    pub direction: Direction,
    pub asset_type: AssetType,
    pub ticker: Option<String>,
    /// As entered — in sats when `bitcoin_unit_is_sats`, otherwise whole units.
    pub quantity: Decimal,
    pub bitcoin_unit_is_sats: bool,
    pub custodian_id: Option<String>,
    pub wallet_id: Option<String>,
}

impl Trade {
    /// Quantity in whole units. Sats are converted here so everything below
    /// this point is in one unit — the conversion is exact in `Decimal`, where
    /// dividing by 100,000,000 loses nothing.
    fn units(&self) -> Decimal {
        if self.asset_type == AssetType::Bitcoin && self.bitcoin_unit_is_sats {
            self.quantity / Decimal::from(SATS_PER_BTC)
        } else {
            self.quantity
        }
    }

    fn holding_type(&self) -> &'static str {
        match self.asset_type {
            AssetType::Stock => "STOCK",
            AssetType::Bitcoin => "BITCOIN",
        }
    }
}

/// Apply a trade to its holding, or reverse it with `multiplier = -1`.
///
/// Returns `Some(cost_basis_allocated)` for a sell — the portion of the
/// holding's basis this sale consumed, which the caller writes back onto the
/// transaction. `None` for a buy, which allocates nothing.
///
/// **The multiplier inverts the direction, it does not negate the amounts.** A
/// BUY reversal behaves exactly like a SELL and vice versa; that is what makes
/// a reversal restore the holding rather than merely subtract from it.
pub async fn apply_trade_to_holding(
    conn: &mut SqliteConnection,
    trade: &Trade,
    usd_amount: Cents,
    multiplier: i8,
) -> Result<Option<Cents>> {
    let quantity = trade.units();
    let holding_type = trade.holding_type();

    let existing = sqlx::query!(
        r#"SELECT "id" AS "id!: String", "quantity" AS "quantity!: String",
                  "costBasis" AS "cost_basis: i64"
           FROM "InvestmentHolding"
           WHERE "type" = ?
             AND ("custodianId" IS ? OR (?  IS NULL AND "custodianId" IS NULL))
             AND ("walletId" IS ? OR (? IS NULL AND "walletId" IS NULL))
             AND ("ticker" IS ?)
           LIMIT 1"#,
        holding_type,
        trade.custodian_id,
        trade.custodian_id,
        trade.wallet_id,
        trade.wallet_id,
        trade.ticker,
    )
    .fetch_optional(&mut *conn)
    .await
    .context("looking up holding")?;

    let is_buy = matches!(
        (trade.direction, multiplier),
        (Direction::Buy, 1) | (Direction::Sell, -1)
    );

    if is_buy {
        match existing {
            Some(h) => {
                // Quantity arithmetic in Rust, deliberately. `SET quantity =
                // quantity + ?` would make SQLite coerce the TEXT to a float.
                let current = Decimal::from_str(&h.quantity).context("holding quantity")?;
                let updated = (current + quantity).normalize().to_string();
                let basis = Cents(h.cost_basis.unwrap_or(0)) + usd_amount;
                sqlx::query!(
                    r#"UPDATE "InvestmentHolding" SET "quantity" = ?, "costBasis" = ? WHERE "id" = ?"#,
                    updated,
                    basis.0,
                    h.id
                )
                .execute(&mut *conn)
                .await?;
            }
            None => {
                let name = entity_name(conn, trade).await?;
                let id = format!("hold_{}", uuid_like(&name, trade));
                let qty = quantity.normalize().to_string();
                sqlx::query!(
                    r#"INSERT INTO "InvestmentHolding"
                         ("id","name","ticker","type","quantity","costBasis",
                          "createdAt","updatedAt","custodianId","walletId")
                       VALUES (?,?,?,?,?,?,datetime('now'),datetime('now'),?,?)"#,
                    id,
                    name,
                    trade.ticker,
                    holding_type,
                    qty,
                    usd_amount.0,
                    trade.custodian_id,
                    trade.wallet_id,
                )
                .execute(&mut *conn)
                .await?;
            }
        }
        return Ok(None);
    }

    // Sell (or a buy being reversed).
    let Some(h) = existing else {
        // Nothing to decrement. Reachable only if validation upstream let a
        // sell through for an asset that was never held.
        return Ok(None);
    };

    let current_qty = Decimal::from_str(&h.quantity).context("holding quantity")?;
    let current_basis = Cents(h.cost_basis.unwrap_or(0));

    // Proportional: (sold / held) × basis. The one division here.
    let proportion = if current_qty > Decimal::ZERO {
        quantity / current_qty
    } else {
        Decimal::ZERO
    };
    let reduction_exact = Decimal::from(current_basis.0) * proportion;
    let reduction = Cents(
        reduction_exact
            .round_dp_with_strategy(0, RoundingStrategy::MidpointAwayFromZero)
            .to_i64()
            .context("cost basis reduction overflows i64 cents")?,
    );

    let new_qty = (current_qty - quantity).normalize().to_string();
    let new_basis = current_basis - reduction;

    sqlx::query!(
        r#"UPDATE "InvestmentHolding" SET "quantity" = ?, "costBasis" = ? WHERE "id" = ?"#,
        new_qty,
        new_basis.0,
        h.id
    )
    .execute(&mut *conn)
    .await?;

    Ok(Some(reduction))
}

/// The holding's display name comes from the custodian or wallet it sits in.
async fn entity_name(conn: &mut SqliteConnection, trade: &Trade) -> Result<String> {
    match trade.asset_type {
        AssetType::Stock => {
            let id = trade
                .custodian_id
                .as_deref()
                .context("stock trade needs a custodian")?;
            let r = sqlx::query!(
                r#"SELECT "name" AS "name!: String" FROM "Custodian" WHERE "id" = ?"#,
                id
            )
            .fetch_one(&mut *conn)
            .await
            .context("custodian not found")?;
            Ok(r.name)
        }
        AssetType::Bitcoin => {
            let id = trade
                .wallet_id
                .as_deref()
                .context("bitcoin trade needs a wallet")?;
            let r = sqlx::query!(
                r#"SELECT "name" AS "name!: String" FROM "Wallet" WHERE "id" = ?"#,
                id
            )
            .fetch_one(&mut *conn)
            .await
            .context("wallet not found")?;
            Ok(r.name)
        }
    }
}

/// A stable id for a new holding. Deterministic rather than random so a test
/// can assert on it; the uniqueness that matters is enforced by the lookup
/// above, which finds any existing holding before one is created.
fn uuid_like(name: &str, trade: &Trade) -> String {
    format!(
        "{}_{}_{}",
        trade.holding_type().to_lowercase(),
        trade.ticker.as_deref().unwrap_or("na"),
        name.to_lowercase().replace(' ', "_")
    )
}

/// Apply a Bitcoin PAYMENT to its wallet holding — port of
/// `applyBitcoinToHolding` (bitcoin-holding hook, priority 10).
///
/// Distinct from a trade: spending or receiving BTC directly, rather than
/// buying or selling it. `BitcoinPaymentDetail` is its own table (ADR-027).
///
/// **It is the same operation as a trade with the direction derived
/// differently**, which is not obvious and is worth the truth table:
///
/// | tx type          | multiplier | effect    | equivalent trade |
/// |------------------|-----------:|-----------|------------------|
/// | EXPENSE          |         +1 | decrement | Sell +1          |
/// | EXPENSE          |         -1 | increment | Sell -1          |
/// | INCOME / REFUND  |         +1 | increment | Buy  +1          |
/// | INCOME / REFUND  |         -1 | decrement | Buy  -1          |
///
/// So spending maps to `Sell` and receiving maps to `Buy`, and the existing
/// multiplier logic then produces the right answer in all four cases. Sharing
/// the implementation means the proportional cost-basis rule and the decimal
/// quantity handling cannot drift between the two hooks — which is the whole
/// reason `classifyLeftovers` was centralised on the reconcile side too.
///
/// Returns the cost basis released on a decrement. The TypeScript discards it
/// (the function returns void); it is surfaced here because the caller may want
/// it and throwing information away is harder to undo than ignoring it.
pub async fn apply_bitcoin_payment_to_holding(
    conn: &mut SqliteConnection,
    wallet_id: &str,
    quantity: Decimal,
    unit_is_sats: bool,
    tx_type: &str,
    usd_amount: Cents,
    multiplier: i8,
) -> Result<Option<Cents>> {
    let direction = match tx_type {
        // Spending BTC reduces the holding, exactly as a sale does.
        "EXPENSE" => Direction::Sell,
        // Receiving BTC increases it.
        "INCOME" | "REFUND" => Direction::Buy,
        other => anyhow::bail!("bitcoin payment on an unsupported transaction type: {other}"),
    };

    let trade = Trade {
        direction,
        asset_type: AssetType::Bitcoin,
        ticker: None,
        quantity,
        bitcoin_unit_is_sats: unit_is_sats,
        custodian_id: None,
        wallet_id: Some(wallet_id.to_string()),
    };

    apply_trade_to_holding(conn, &trade, usd_amount, multiplier).await
}
